docs/design/002-containerized-gadgets.md

```diff
@@ -0,0 +1,685 @@
+# Containerized Gadgets Support Inspektor Gadget
+
+The gadgets we provide are heavily coupled to Inspektor Gadget as they are all developed on the same
+repository and maintained by our team. Adding new gadgets implies that they have to be reviewed and
+approved by us. This approach presents some disadvantages:
+- Users need to install new versions of Inspektor Gadget to get new gadgets (or updates to them).
+- Too specific gadgets can't be added to our code base.
+- Users can't have private gadgets.
+
+The [containerized gadgets idea](https://github.com/inspektor-gadget/inspektor-gadget/issues/1669)
+aims to make Inspektor Gadget a framework to run eBPF programs (gadgets), like Docker does for
+containers. This approach decouples the gadgets from Inspektor Gadget solving the issues mentioned
+above.
+
+Many ideas of this design document are based on [bumblebee](https://github.com/solo-io/bumblebee)
+and [ebpf_exporter](https://github.com/cloudflare/ebpf_exporter) projects.
+
+## User Experience
+
+The UX we provide should be very close to the existing container runtimes like Docker. This document
+considers the client-server setup for ig described in
+https://github.com/inspektor-gadget/inspektor-gadget/issues/1681. It's being implemented at the time
+of writing this document.
+
+### `run`
+
+The run command starts a gadget:
+
+```bash
+$ gadgetctl run mygadgetimage:latest
+# print output to terminal
+```
+
+`mygadgetimage:latest` is a gadget OCI image.
+This command should provide a flag to run on detached mode:
+
+```bash
+$ gadgetctl run mygadgetimage:latest --detach
+a3397355fc4b2ba49cdc1ab2d36f728cbda65f0ef979854186f2ea780e8659e1
+```
+
+This command automatically pulls the image if it is not available in the local image cache.
+
+### `list`
+
+List print the running gadgets
+
+```bash
+$ gadgetctl list
+...
+```
+
+### `attach`
+
+Attach to a running gadget
+
+```bash
+$ gadgetctl attach mygadget
+# gets terminal attached to the gadget
+```
+
+### `pull`
+
+Pulls a gadget image
+
+```bash
+$ gadgetctl pull mygadgetimage:latest
+```
+
+### `build`
+
+Builds and packages a gadget as an OCI image
+
+```bash
+$ gadgetctl build --prog foo.bpf.c --definition foo.yaml mygadgetimage:latest
+```
+
+Details of this command are still to be defined. Like other flags that are supported.
+
+### `tag`
+
+Tag an image
+
+```bash
+$ gadgetctl tag mygadgetimage:latest ghcr.io/foo/mygadgetimage:v1
+```
+
+### `image`
+
+Different commands to handle images
+
+- `image list`
+- `image delete`
+etc...
+
+TODO: Should commands like `ig pull`, `ig build`, `ig push` be aliases of `ig image ...` ones?
+
+## Gadget Packaging
+
+Gadgets should be packed as [OCI](https://github.com/opencontainers/image-spec) images. Initially
+the following layers are proposed:
+- `application/config.image.inspektor-gadget.io+yaml`: [Configuration file](#configuration-file)
+- `application/ebpf.image.inspektor-gadget.io+binary`: Compiled eBPF object.
```

Comment:
I think you're right, we should use the [vendor tree](https://en.wikipedia.org/wiki/Media_type#Vendor_tree) because:

> Registrations in the standards tree must be either associated with IETF specifications approved directly by the IESG, or registered by an IANA recognized standards-related organization

And we are not registering our media-type with any standards-related organization. So it should be namespaced correctly.

https://datatracker.ietf.org/doc/html/rfc2048#section-2.1

>    Registrations in the vendor tree will be distinguished by the leading
   facet "vnd.".  That may be followed, at the discretion of the
   registration, by either a media type name from a well-known producer
   (e.g., "vnd.mudpie") or by an IANA-approved designation of the
   producer's name which is then followed by a media type or product
   designation (e.g., vnd.bigcompany.funnypictures).

Reply:
Actually I have a question here. We discussed to use something different in https://github.com/inspektor-gadget/inspektor-gadget/pull/1863/files#r1273752216, which one should we use? It seems like there is not a standard one for ebpf ELF files.